manager/manager.py

```diff
@@ -0,0 +1,1105 @@
+        :param upload_or_download: "upload" or "download"
+        :param preview:  see upload_project_folder_or_file*(
+        """
+        local_filepath = self._join("local", filepath)
+        remote_filepath = self._join("remote", filepath)
+
+        local = FsTarget(local_filepath)
+        remote = self._get_remote_target(remote_filepath)
+
+        opts = self._get_default_syncronizer_opts(preview)
+
+        syncronizer = self._get_syncronizer(upload_or_download)
+        s = syncronizer(local, remote, opts)
+        s.run()
+
+    def _get_syncronizer(self, upload_or_download: str):
+        """
+        Convenience function to get the pyftpsync syncronizer
+        """
+        if upload_or_download == "upload":
+            syncronizer = UploadSynchronizer
+
+        elif upload_or_download == "download":
+            syncronizer = DownloadSynchronizer
+
+        return syncronizer
+
+    def _get_remote_target(self, remote_filepath: str):
+        """
+        Convenience function to get the pyftsync target
+        based on remote connection type.
+        """
+        if self.cfg["ssh_to_remote"]:
+
+            remote = SFTPTarget(
+                remote_filepath,
+                self.cfg["remote_host_id"],
+                username=self.cfg["remote_host_username"],
+                private_key=self._ssh_key_path,
+                hostkeys=self._hostkeys,
+            )
+
+        else:
+            remote = FsTarget(remote_filepath)
+
+        return remote
+
+    def _get_default_syncronizer_opts(self, preview: bool):
+        """
+        Retrieve the default options for upload and download. These
+        are very important as define the behaviour of file transfer
+        when there are conflicts (e.g. whether to delete remote
+        file if it is not found on the local filesystem).
+
+        Currently, all options are set so that no file is ever overwritten.
+        If there is a remote folder that is older than the local folder, it will not
+        be overwritten. The only 'overwrite' that occurs is if the remote
+        or local folder has been deleted - by default this will not be replaced as
+        pyftpsync metadata indicates the file has been deleted. Using the default
+        'force' option will force file transfer, but also has other effects e.g.
+        overwriting newer files with old, which we dont want. This option has been
+        edited to permit a "restore" argumnent, which acts Force=False except
+        in the case where the local / remote file has been deleted entirely, in which
+        case it will be replaced.
+
+        :param preview: run pyftpsync's "dry_run" option.
+        """
+        opts = {
+            "help": False,
+            "verbose": 5,
+            "quiet": 0,
+            "debug ": False,
+            "case": "strict",
+            "dry_run": preview,
+            "progress": False,
+            "no_color": True,
+            "ftp_active": False,
```

Comment:
Will these ever need to be changed, i.e. should they be hardcoded?

Reply:
I'm not certain, currently they are set to be as un-intrusive as possible (e.g. never overwrite an existing file). In theory a user might want to edit them, and during this testing phase we might make amendments. But generally, there will probably be a default set that is strongly advised to use.

Reply:
yes I think so, and it would be nice to have at least a subset of these as possible options. I'm testing at the moment and will check if the pyftpsync provides functionality to deal with disrupted connections etc. If not, it is probably worth jettisoning this and using rsync / robocopy depending on os.